contracts/btc-finality/src/liveness.rs

```diff
@@ -1,37 +1,60 @@
 use crate::{
     error::ContractError,
+    events::{new_finality_provider_status_change_event, FinalityProviderStatus},
     state::config::Config,
     state::finality::{get_last_signed_height, get_power_table_at_height, JAIL},
 };
-use cosmwasm_std::{DepsMut, Env};
+use cosmwasm_std::{DepsMut, Env, Event};
 
 /// Check for inactive finality providers, and jail them.
 ///
 /// The jailing mechanism is meant to be simple where an active finality provider
 /// is jailed if it has not signed a block in the last `missed_blocks_window` blocks.
 ///
 /// Note that this takes effect only after the next block is processed.
-pub fn handle_liveness(deps: &mut DepsMut, env: &Env, cfg: &Config) -> Result<(), ContractError> {
+/// Returns events for newly jailed finality providers.
+pub fn handle_liveness(
+    deps: &mut DepsMut,
+    env: &Env,
+    cfg: &Config,
+) -> Result<Vec<Event>, ContractError> {
     let fp_power_table = get_power_table_at_height(deps.storage, env.block.height)?;
     let window_start_height = env.block.height.saturating_sub(cfg.missed_blocks_window);
 
+    let mut events = Vec::new();
+
     for (fp_btc_pk_hex, _) in fp_power_table {
         let last_sign_height = get_last_signed_height(deps.storage, &fp_btc_pk_hex)?;
         let inactive = match last_sign_height {
             Some(h) if h > window_start_height => false,
             _ => true,
         };
         if inactive {
-            // Jail if not already jailed.
-            JAIL.update(deps.storage, &fp_btc_pk_hex, |existing| {
-                Ok::<_, ContractError>(
-                    existing.unwrap_or(env.block.time.seconds() + cfg.jail_duration),
-                )
-            })?;
+            // Check if FP is already jailed to avoid duplicate events and unnecessary updates
+            let was_already_jailed = JAIL.has(deps.storage, &fp_btc_pk_hex);
+
+            if was_already_jailed {
+                // FP is already jailed, no action needed
+                continue;
+            }
```

Comment:
```suggestion
            // Check if FP is already jailed to avoid duplicate events and unnecessary updates
            if JAIL.has(deps.storage, &fp_btc_pk_hex) {
                // FP is already jailed, no action needed
                continue;
            }
```